x/ratelimit/keeper/rate_limit.go

```diff
@@ -0,0 +1,198 @@
+package keeper
+
+import (
+	"strings"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/Stride-Labs/stride/v5/x/ratelimit/types"
+)
+
+// Get the rate limit byte key built from the denom and channelId
+func GetRateLimitItemKey(denom string, channelId string) []byte {
+	return append(types.KeyPrefix(denom), types.KeyPrefix(channelId)...)
+}
+
+// The total value on a given path (aka, the denominator in the percentage calculation)
+// is the total supply of the given denom
+func (k Keeper) GetChannelValue(ctx sdk.Context, denom string) sdkmath.Int {
+	return k.bankKeeper.GetSupply(ctx, denom).Amount
+}
+
+// If the rate limit is exceeded or the denom is blacklisted, we emit an event
+func EmitRateLimitExceededEvent(ctx sdk.Context, denom, channelId string, direction types.PacketDirection, amount sdkmath.Int, err error) {
```

Comment:
Nit: might be nice to have separate events for attempts to transfer blacklisted assets vs attempts to transfer assets whose rate limits were exceeded. The former seems important to be able to identify at a moment's notice in potential emergency scenarios

Reply:
I went back and forth on that a little. My thinking was you could easily determine whether it was blacklisted or throttled based on the error message and I thought it would be nice to have a common event type to listen to all blocked txs.

Another option is to change the event type to TransferBlocked and then add an attribute that's either RateLimitExceeded or BlacklistedDenom? 
Also fine with just using different event types

Wdyt?

Reply:
> change the event type to TransferBlocked and then add an attribute that's either RateLimitExceeded or BlacklistedDenom?

I think this would be nice n clean. I know this is nit, but it could make a big difference in the moment if there's ever a blacklist event